lib/tbot/tbot.go

```diff
@@ -569,3 +575,40 @@ func (p *proxyPingCache) ping(ctx context.Context) (*webclient.PingResponse, err
 
 	return p.cachedValue, nil
 }
+
+type alpnConnUpgradeParams struct {
+	addr     string
+	insecure bool
+}
+
+type alpnProxyConnUpgradeRequiredCache struct {
+	botCfg *config.BotConfig
+	log    *slog.Logger
+
+	mu    sync.RWMutex
+	cache map[alpnConnUpgradeParams]bool
+}
+
+func (a *alpnProxyConnUpgradeRequiredCache) isUpgradeRequired(ctx context.Context, addr string, insecure bool) (bool, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
```

Comment:
Can we use a `golang.org/x/sync/singleflight.Group` here (and a map with a regular mutex)? As written, this serializes checks for different addresses (and `insecure`s).

Reply:
Sidenote: why isn't there a singleflight-like construct that also keeps the data until discarded? 😢 

Also, why isn't there a generic singleflight? 😭

Reply:
02e185c550